src/main/java/org/spongepowered/mod/service/scheduler/AsyncScheduler.java

```diff
@@ -0,0 +1,735 @@
+                }
+            }
+        }
+    }
+
+    /**
+     * <p>Returns the instance (handle) to the Asynchronous TaskScheduler.</p>
+     *
+     * <p>
+     * A static reference to the TaskScheduler singleton is returned by
+     * the function getInstance().  The implementation of getInstance follows the usage
+     * of the AtomicReference idiom.</p>
+     *
+     * @return The single interface to the Asynchronous Scheduler
+     */
+    /**
+     * <p>Returns the instance (handle) to the Synchronous TaskScheduler.</p>
+     * <p/>
+     * <p>
+     * A static reference to the Synchronous Scheduler singleton is returned by
+     * the function getInstance().  The implementation of getInstance follows the usage
+     * of the AtomicReference idiom.</p>
+     *
+     * @return The single interface to the Synchronous Scheduler
+     */
+
+    private static class AsynchronousSchedulerSingletonHolder {
+        private static final AsynchronousScheduler INSTANCE = new AsyncScheduler();
+    }
+
+    public static AsynchronousScheduler getInstance() {
+        return AsynchronousSchedulerSingletonHolder.INSTANCE;
+    }
+
+    private void recalibrateMinimumTimeout() {
+        lock.lock();
+        try {
+            for (ScheduledTask task : this.taskList) {
+                // Tasks may have dropped off the list.
+                // Whatever Tasks remain, recalibrate:
+                //
+                // Recalibrate the wait delay for processing tasks before new tasks
+                // cause the scheduler to process pending tasks.
+                minimumTimeout = Long.MAX_VALUE;
+                if (task.offset == 0 && task.period == 0) {
+                    minimumTimeout = 0;
+                } else if (task.offset > 0 && task.period == 0) {
+                    // task with non-zero offset, zero period
+                    minimumTimeout = Math.min(task.offset, minimumTimeout);
+                } else if (task.offset == 0 && task.period > 0) {
+                    // task with zero offset, non-zero period
+                    minimumTimeout = Math.min(task.period, minimumTimeout);
+                } else if (task.offset > 0 && task.period > 0) {
+                    // task with non-zero offset, non-zero period
+                    minimumTimeout = Math.min(task.offset, minimumTimeout);
+                    minimumTimeout = Math.min(task.period, minimumTimeout);
+                }
+            }
+
+            // If no tasks remain, recalibrate to max timeout
+            if (taskList.isEmpty()) {
+                minimumTimeout = Long.MAX_VALUE;
+            }  else {
+                long latency = System.currentTimeMillis() - lastProcessingTimestamp;
+                minimumTimeout -= (latency <= 0) ? 0 : latency;
+                minimumTimeout = (minimumTimeout < 0) ? 0 : minimumTimeout;
+            }
+        } finally {
+            lock.unlock();
+        }
+    }
+
+    private void ProcessTasks() {
```

Comment:
note sure where you're indicating.

taskList ?
latency ?
lastProcesingTimestamp ?
lock ?